central/processlisteningonport/datastore/datastore_impl.go

```diff
@@ -0,0 +1,166 @@
+package datastore
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/stackrox/rox/central/metrics"
+	processIndicatorStore "github.com/stackrox/rox/central/processindicator/datastore"
+	"github.com/stackrox/rox/central/processlisteningonport/store/postgres"
+	"github.com/stackrox/rox/central/role/resources"
+	v1 "github.com/stackrox/rox/generated/api/v1"
+	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/logging"
+	"github.com/stackrox/rox/pkg/sac"
+	"github.com/stackrox/rox/pkg/search"
+	"github.com/stackrox/rox/pkg/uuid"
+)
+
+type datastoreImpl struct {
+	storage        postgres.Store
+	indicatorStore processIndicatorStore.DataStore
+}
+
+var (
+	plopSAC = sac.ForResource(resources.ProcessListeningOnPort)
+	log     = logging.LoggerForModule()
+)
+
+func newDatastoreImpl(
+	storage postgres.Store,
+	indicatorDataStore processIndicatorStore.DataStore,
+) *datastoreImpl {
+	return &datastoreImpl{
+		storage:        storage,
+		indicatorStore: indicatorDataStore,
+	}
+}
+
+func (ds *datastoreImpl) AddProcessListeningOnPort(
+	ctx context.Context,
+	portProcesses ...*storage.ProcessListeningOnPort,
+) error {
+	defer metrics.SetDatastoreFunctionDuration(
+		time.Now(),
+		"ProcessListeningOnPort",
+		"AddProcessListeningOnPort",
+	)
+
+	var (
+		lookups       []*v1.Query
+		indicatorsMap = map[string]*storage.ProcessIndicator{}
+	)
+
+	if ok, err := plopSAC.WriteAllowed(ctx); err != nil {
+		return err
+	} else if !ok {
+		return sac.ErrResourceAccessDenied
+	}
+
+	// First query all needed process indicators in one go
+	for _, val := range portProcesses {
+		lookups = append(lookups,
+			search.NewQueryBuilder().
+				AddExactMatches(search.ContainerName, val.Process.ContainerName).
```

Comment:
is there any chance `val.Process` can be nil?  May want to protect ourselves for that case.

Reply:
Probably no, but it never hurts to check.